### ISSUE TITLE
Rate control

### ISSUE THREAD

**Issue by jnoring** [w3c/mediacapture-record#57]
I was happy to see the addition of controls to specify bitrate in https://github.com/w3c/mediacapture-record/issues/20, but coming from a video background, it's still pretty limiting.  I'd like to discuss other options for incorporating rate control into the media capture API.  Is this the appropriate place to have that discussion?

Thanks!

**Comment by jnoring** [comment]
This might also be a bigger discussion of how to abstract an interface to all of the encoders MediaRecorder potentially uses to expose more fine-grained encoder settings (for example, rate control, keyframe insertion, profiles, quality settings, etc.) without exposing javascript developers to the craziness that is https://trac.ffmpeg.org/wiki/Encode/VP8 and https://trac.ffmpeg.org/wiki/Encode/H.264.  Has there been any discussion of that?

**Comment by yell0wd0g** [comment]
@jnoring, yes this is a good place to start the discussion. By preference with some kind of proposal that Gecko and Chrome can react to (Note that H264 is also coming to Chrome: https://crbug.com/601636).

**Comment by dontcallmedom** [comment]
hi @jnoring , before you start making suggestions for substantive changes to the spec, I want to make sure you're aware of the [requirements for contributing](https://github.com/w3c/mediacapture-record/blob/master/CONTRIBUTING.md) to W3C specifications.

In particular, it would be useful to know your corporate affiliation (if any) to understand better how we can accept your contributions while keeping the spec we develop in line with the W3C royalty-free policy — feel free to contact me at dom@w3.org if you want to discuss this in a less public forum :)

**Comment by jnoring** [comment]
@dontcallmedom thank you for sharing the contribution requirements.  I'll figure that out with my employer (HireVue); I'll email you regarding that.  Thanks!

**Comment by yell0wd0g** [comment]
@jnoring do you have any update...?

**Comment by yell0wd0g** [comment]
@jnoring, ping 2, I'd love to hear your contribution(s).

**Comment by jnoring** [comment]
At the risk of making something complicated, I'd really like a "quality" setting (which implies a variable bitrate) in addition to the hard-coded bitrate.

for example, with x264, you can configure it with a bitrate target (constant bitrate) or a quality target (crf) where bitrate is allowed to float to achieve consistent quality as opposed to consistent bitrate.  Also, there are hybrid modes with a quality target and bitrate cap (so maybe `-crf 23`, max bitrate: 1500 kbps).

What I'm not sure about is how to really abstract this out.  For example, in my usage of media recorder, I'd really like a consistent _quality_ to the video.  Right now I use a bitrate target to keep the files from being gigantic, but in the perfect world I'd just do something akin to `-crf 23` and let the bitrate only be as high as the encoder decides it needs to be.

**Comment by yell0wd0g** [comment]
Could that (the semantics of the bitrate parameter) be conveyed in [MediaRecorderOptions](https://w3c.github.io/mediacapture-record/MediaRecorder.html#MediaRecorderOptions)? Perhaps as another parameter (for [cbr/vbr/ rq](http://www.webmproject.org/docs/encoder-parameters/#vbr-cbr-and-cq-mode), or  [crf](http://slhck.info/articles/crf))?

The constant-quality target might be beyond the capabilities of this spec :-)

**Comment by mikeal** [comment]
Additionally, we really need to be able to adjust the time delay between keyframes.

As it stands in Chrome, audio-only is 30s and video is 3s. This means that audio-only is effectively unusable for live broadcasting. In addition to rate control I'd like to see some keyframe settings.

**Comment by jnoring** [comment]
@miguelao honestly, I'm not sure of a clean way to abstract this because it's ultimately highly dependent on the encoder implementation, as well as individual browser use of those encoders.  To get clear, consistent results between implementations would be difficult.

I'll think about it more, but for the time being, I'd close it.  @mikeal 's suggestion is a good one though.

**Comment by mikeal** [comment]
@jnoring because there are no required encoders in the spec people are getting by with [MediaRecorder.isTypeSupported](https://developer.mozilla.org/en-US/docs/Web/API/MediaRecorder/isTypeSupported). The problem you describe is similar, that each encoder has differing options and there's no way to abstract it. The way the browser has dealt with this so far is to give this information to the user and to not support calling these methods with unsupported options.

I would recommend something like a `MediaRecorder.supportedMedia()`method which would return information about the supported content types and their encoding options. If the user tries to use an unsupported option then it would be perfectly acceptable just to throw.

In other words, rather than trying to abstract differing encoder options, which is probably impossible, let's just bubble this information up to the user and throw if they use unsupported options.

**Comment by friday** [comment]
Constant quality would be really useful. Given the option, I don't see why anyone would want to use cbr or vbr rather than constant quality for MediaRecorder.

CRF is supported by h264, libx264 and h265 (I don't think any browser can record with h265 currently, but that could change if Safari implements MediaRecorder). ffmpeg also supports crf for vp8 and vp9 (unlike libvpx/vpxenc). I don't know about av1, but I'm guessing the arguments will be similar to vp9.

As has been pointed out CRF is not consistent between codecs:
* CRF in h264 and h265 uses a scale of 0-51 for 8-bit and 0-63 for 10 bit. It's exponential, and "increasing the CRF value +6 is roughly half the bitrate while -6 is roughly twice the bitrate" ([source](https://trac.ffmpeg.org/wiki/Encode/H.264#crf)).
* For VP8 and VP9 the scale is 0-63, though VP8 can't use 0-3 ([source](https://trac.ffmpeg.org/wiki/Encode/VP8#VariableBitrate)). I couldn't find much information about it. It seems to me that "CRF" in ffmpeg uses libvpx's`--cq-level` underneath.
* Rates in h264 and h265 correspond to bitrates rather than quality. "The CRF of 28 [encoding with h265] should visually correspond to libx264 video at CRF 23" ([source](https://trac.ffmpeg.org/wiki/Encode/H.265#ConstantRateFactorCRF)). I would expect VP8/VP9 to differ as well.

All of this makes it impossible to abstract CRF into a consistent cross browser standard without doing something ridiculous like standardizing and maintaining implementation details for each codec. But at least that's one less aspect to discuss.

Spec suggestions:

* Add 1 new quality parameter to [MediaRecorderOptions](https://w3c.github.io/mediacapture-record/MediaRecorder.html#mediarecorderoptions-section)
* Since the scope of the quality rate we're discussing applies to video only, the parameter name should start with "video", consistently with "videoBitsPerSecond". Ex "videoQualityRate".
* "videoQualityRate" should map to the codec quality rate parameter (ex: crf or cq-level)
* If specifying both "videoQualityRate" and (video)bitsPerSecond", bitsPerSecond should act as a fallback for codecs (and browsers) lacking support for quality rate.
* Throw a helpful error if specifying videoQualityRate in [MediaRecorderOptions](https://w3c.github.io/mediacapture-record/MediaRecorder.html#mediarecorderoptions-section), but codec is missing from mimeType (to avoid unexpected surprises in other browsers and/or when browsers change default codecs)

(I have no corporate affiliations and my suggestions are royalty-free)

**Comment by trianglestrip** [comment]
any update of this?

**Comment by Queatz** [comment]
Still waiting

**Comment by pkunszt** [comment]
Did anything happen here?
